File: 1549-longest-continuous-subarray-with-absolute-diff-less-than-or-equal-to-limit/longest-continuous-subarray-with-absolute-diff-less-than-or-equal-to-limit.cpp
class Solution {
public:
    int longestSubarray(vector<int>& n, int li) {
        deque<int> i, d;
        int l = 0, m = 0;
        for (int r = 0; r < n.size(); r++) {
 
            while (!i.empty() && n[r] < i.back()) i.pop_back();
            i.push_back(n[r]);
            
            while (!d.empty() && n[r] > d.back()) d.pop_back();
            d.push_back(n[r]);
            while (d.front() - i.front() > li) {
                if (n[l] == d.front()) d.pop_front();
                if (n[l] == i.front()) i.pop_front();
                l++;
            }

            m = max(m, r - l + 1);
        }
        return m;
    }
};

File: 1549-longest-continuous-subarray-with-absolute-diff-less-than-or-equal-to-limit/longest-continuous-subarray-with-absolute-diff-less-than-or-equal-to-limit.go
func longestSubarray(n []int, li int) int {
    i := list.New() 
    d := list.New() 
    m, l := 0, 0

    for r := 0; r < len(n); r++ {
 
        for i.Len() > 0 && n[r] < i.Back().Value.(int) {
            i.Remove(i.Back())
        }
        i.PushBack(n[r])

        for d.Len() > 0 && n[r] > d.Back().Value.(int) {
            d.Remove(d.Back())
        }
        d.PushBack(n[r])

        for d.Front().Value.(int)-i.Front().Value.(int) > li {
            if n[l] == d.Front().Value.(int) {
                d.Remove(d.Front())
            }
            if n[l] == i.Front().Value.(int) {
                i.Remove(i.Front())
            }
            l++
        }

        if m < r-l+1 {
            m = r - l + 1
        }
    }

    return m
}
